sktime/decomposition/vampnet.py

```diff
@@ -0,0 +1,662 @@
+    is implemented in :meth:`score`."""
+    return -1. * score(data, data_lagged, method=method, epsilon=epsilon, mode=mode)
+
+
+class MLPLobe(nn.Module):
+    r""" A multilayer perceptron which can be used as a neural network lobe for VAMPNets. """
+
+    def __init__(self, units: List[int], nonlinearity=nn.ELU, initial_batchnorm: bool = True,
+                 output_nonlinearity=lambda: nn.Softmax(dim=1)):
+        r"""Instantiates a new lobe.
+
+        Parameters
+        ----------
+        units : list of integers
+            The units of the fully connected layers.
+        nonlinearity : callable, default=torch.nn.ELU
+            A callable (like a constructor) which yields an instance of a particular activation function.
+        initial_batchnorm : bool, default=True
+            Whether to use batch normalization before the data enters the rest of the network.
+        output_nonlinearity : callable, default=softmax
+            The output activation/nonlinearity. If the data decomposes into states, it can make sense to use
+            an output activation like softmax which produces a probability distribution over said states.
+        """
+        super().__init__()
+        layers = []
+        if initial_batchnorm:
+            layers.append(nn.BatchNorm1d(units[0]))
+        for fan_in, fan_out in zip(units[:-2], units[1:-1]):
+            layers.append(nn.Linear(fan_in, fan_out))
+            layers.append(nonlinearity())
+        layers.append(nn.Linear(units[-2], units[-1]))
+        if output_nonlinearity is not None:
+            layers.append(output_nonlinearity())
+        self._sequential = nn.Sequential(*layers)
+
+    def forward(self, inputs):
+        return self._sequential(inputs)
+
+
+class VAMPNetModel(Model, Transformer):
+    r"""
+    A VAMPNet model which can be fit to data optimizing for one of the implemented VAMP scores.
+
+    See Also
+    --------
+    VAMPNet : The corresponding estimator.
+    """
+
+    def __init__(self, lobe: nn.Module, lobe_timelagged: Optional[nn.Module] = None,
+                 dtype=np.float32, device=None):
+        r"""Creates a new VAMPNet estimator instance.
+
+        Parameters
+        ----------
+        lobe : torch.nn.Module
+            One of the lobes of the VAMPNet. See also :class:`MLPLobe`.
+        lobe_timelagged : torch.nn.Module, optional, default=None
+            The timelagged lobe. Can be left None, in which case the lobes are shared.
+        dtype : data type, default=np.float32
+            The data type for which operations should be performed. Leads to an appropriate cast within fit and
+            transform methods.
+        device : device, default=None
+            The device for the lobe(s). Can be None which defaults to CPU.
+        """
+        super().__init__()
+        self._lobe = lobe
+        self._lobe_timelagged = lobe_timelagged if lobe_timelagged is not None else lobe
```

Comment:
i do not know the answer to this but does it need to be a hard copy? do they share weights/training and stuff? just thinking of how this interfaces with lines 314-5

Reply:
could be a deep copy but usually it is not from what I understand